oarepo_vocabularies/ui/theme/assets/semantic-ui/js/oarepo_vocabularies_ui/form/components/VocabularyTreeSelectField/VocabularyTreeSelectField.jsx

```diff
@@ -0,0 +1,476 @@
+                    onChange={(e) => {
+                      handleSelect(option, e);
+                    }}
+                  />
+                )}
+                <Button
+                  basic
+                  color="black"
+                  onClick={openHierarchyNode(option.value, index)}
+                  onDoubleClick={(e) => {
+                    handleSelect(option, e);
+                  }}
+                  onKeyDown={(e) => {
+                    handleKey(e, index);
+                  }}
+                  tabIndex={0}
+                >
+                  {option.hierarchy.title[0]}
+                </Button>
+                {option.element_type == "parent" && (
+                  <Button onClick={openHierarchyNode(option.value, index)}>
+                    {index !== columnsCount - 1 && (
+                      <Icon name="angle right" color="black" />
+                    )}
+                  </Button>
+                )}
+              </Grid.Row>
+            );
+          }
+        })}
+      </Grid.Column>
+    );
+  };
+
+  return (
+    <React.Fragment>
+      <SelectField
+        selectOnBlur={false}
+        optimized={optimized}
+        onBlur={() => setFieldTouched(fieldPath)}
+        deburr
+        fieldPath={fieldPath}
+        multiple={multiple}
+        featured={featuredOptions}
+        options={serializedOptions}
+        onClick={(e) => handleOpen(e)}
+        onChange={handleChange}
+        value={multiple ? value.map((o) => o?.id) : value?.id}
+        {...uiProps}
+      />
+      <label className="helptext">{helpText}</label>
+
+      {openState && (
+        <Modal
+          onClose={() => setOpenState(false)}
+          onOpen={() => setOpenState(true)}
+          open={openState}
+          className="tree-field"
+        >
+          <ModalHeader>
+            <Grid.Row>
+              <Header as="h3">{placeholder || "Choose Items"}</Header>
+              <Grid.Column>
+                <Input
+                  type="text"
+                  value={query}
+                  onChange={(e) => setQuery(e.target.value)}
+                  placeholder="Search..."
+                />
+              </Grid.Column>
+            </Grid.Row>
+          </ModalHeader>
+
+          <ModalContent>
+            <Grid>
+              <div className="columns-container">
+                <Grid columns={1}>
+                  <Container>
+                    {hierarchicalData.map((column, level) => (
+                      <React.Fragment key={level}>
+                        {renderColumn(column, level)}
+                      </React.Fragment>
+                    ))}
+                  </Container>
+                </Grid>
+              </div>
+            </Grid>
+          </ModalContent>
+          <ModalActions>
+            <Grid.Row className="gapped">
+              <Grid.Row className="gapped">
+                {selectedState.map((i, index) => (
+                  <Label key={index}>
```

Comment:
Do not use Array index in keys, especially if state can change, items can change order, dissapear etc...